packages/financial-templates-lib/src/price-feed/USPACPriceFeed.ts

```diff
@@ -0,0 +1,251 @@
+    return this.currentPrice;
+  }
+
+  public async getHistoricalPrice(time: number, ancillaryData: string, verbose = false): Promise<BN> {
+    // Construct URL.
+    // https://rapidapi.com/principalapis/api/stock-data-yahoo-finance-alternative/
+    const symbolsStr = this.symbols.join("%2C");
+    const url = `https://stock-data-yahoo-finance-alternative.p.rapidapi.com/v8/finance/spark?symbols=${symbolsStr}&range=3mo&interval=1d`;
+    const options = {
+      method: "GET",
+      headers: {
+        "x-rapidapi-host": "stock-data-yahoo-finance-alternative.p.rapidapi.com",
+        "x-rapidapi-key": this.rapidApiKey,
+      },
+    };
+    const response = (await this.networker.getJson(url, options)) as RapidAPIHistoryReponse | null | undefined;
+
+    if (response == null) {
+      throw new Error(`🚨Invalid response from url ${url}`);
+    }
+
+    if (verbose) {
+      console.log(`Rapid API response for url ${url} is ${JSON.stringify(response, null, 2)}`);
+    }
+
+    // Validate response
+    for (const symbol of this.symbols) {
+      const stockData = response[symbol];
+      if (stockData == null) {
+        throw new Error(`🚨Could not parse price result from url ${url}: missing data for symbol ${symbol}`);
+      }
+      if (stockData.timestamp.length == 0 || stockData.timestamp.length != stockData.close.length) {
+        throw new Error(`🚨Could not parse price result from url ${url}: invalid data for symbol ${symbol}`);
+      }
+      for (const ts of stockData.timestamp) {
+        if (typeof ts != "number") {
+          throw new Error("Could not parse data: invalid timestamp type");
+        }
+      }
+      for (const p of stockData.close) {
+        if (typeof p != "number") {
+          throw new Error("Could not parse data: invalid price type");
+        }
+      }
+      for (let i = 1; i < stockData.timestamp.length; i++) {
+        if (stockData.timestamp[i] <= stockData.timestamp[i - 1]) {
+          throw new Error(`Could not parse data: timestamps for ${symbol} are not increasing monotonously`);
+        }
+      }
+    }
+
+    // Get prices
+    const prices = [];
+    for (const symbol of this.symbols) {
+      const stockData = response[symbol];
+      /*
+        if time is greater than last timestamp (that is equal to the beginning
+        of the last market open datetime) plus 24hours then new day begin and
+        we should use current price
+      */
+      if (time >= stockData.timestamp[stockData.timestamp.length - 1] + 3600 * 24) {
+        if (this.currentPrice == null) {
+          throw new Error("currentPrice is null, price feed was not updated");
+        }
+        return this.currentPrice;
+      }
+
+      // iterate over timestamps and prices backwards, until we find timestamp
+      // that is earlier then given time
+      const result = lodash
+        .zip(stockData.timestamp, stockData.close)
+        .reverse()
+        .find((item) => time >= (item[0] as number));
```

Comment:
I understand the returned `stockData.timestamp` is market open timestamp with close price for the same day returned as `stockData.close` - so if price request is for timestamp before closing this logic would return the closing price that is in the future relative to the request timestamp. Hence you would need to convert `stockData.timestamp` to close timestamp so that it can be accurately compared with request timestamp.

Also note that converting from open to close timestamp sometimes would be challenging due to having earlier closing in some of holidays. Hence alternative would be to consider intra-day price data.

Reply:
I reworked to use intraday price. Are any of your concerns still valid?

Reply:
When looking into returned data it seems that also for intraday granularity rapidapi reports 1m period opening timestamps paired with close prices for the same 1m period. This means that requesting price at 10:00:30 AM the above logic normally would match it to 10:00:00 opening price while the corresponding close price would be at 10:01:00 that is after the request timestamp. 

There is an exception to above statement if requesting price during trading hours that is just couple of seconds behind real-time price - for example, if at 10:00:45 AM one would request price at 10:00:30 then the above algorithm would correctly resolve to the closing price for the 9:59:00-10:00:00 period since 10:00:00-10:01:00 period is not yet available.

Reply:
> This means that requesting price at 10:00:30 AM the *above logic* normally would match it to 10:00:00 *opening* price

What _above logic_ do you mean? If you mean the code above, then requesting price for 10:00:30 AM, the code above will find 10:00:00 timestamp and corresponding price, which is, as you say, is *close* price for 10:00:00 - 10:01:00 interval. So the code above will give the *closing* price for the interval, not *opening* price.

Do you think that the correct price for the middle of the interval is the *open* price for the interval, and thus the code above incorrectly returns *close* price for the interval?

Reply:
The real price for the middle of the interval is somewhere between open price and close price. So neither open price nor the close price are absolutely correct. So we have no reasons to prefer one to another.

Since API gives us only close prices, adjusting timestamps trying to get open price will be difficult. We will have to treat differently timestamps that are in the middle of trading day and timestamps that are after the end of trading day, because in the first case we have to get open price, and in the second case we still have to get the closing price for the last interval (15:59:00-16:00:00)

So I think the correct decision would be to use *closing* price for the interval, as it is currently implemented

Reply:
A different way of looking at it is that the price identifier could get the _most recent_ price, which would be an open price in the middle of the day and a closing price after the end of the trading day. I could see the argument that the user would expect the most recent price when making a price request, since the settlement price would not be known at the time of the request.

Another wrinkle is that a bot that responds to price requests automatically might not be able to know how to respond to this price request if it always uses the closing price, since the closing price might not be known yet. In other words, the price request might not be resolvable, since it would often be requesting information that is (very slightly) in the future.

Reply:
> Do you think that the correct price for the middle of the interval is the _open_ price for the interval, and thus the code above incorrectly returns _close_ price for the interval?

In case the liquidation bot requests price within the same 1 minute interval as the current time (e.g. requesting price for timestamp 10:00:30 AM at 10:00:45 AM)  there would not yet be the close price available for the current 1 minute interval. Instead rapidapi would return the close price of the previous interval (ending 10:00:00 AM). This could become problematic if the dispute bot uses this price feed a bit later (e.g. requesting price for timestamp 10:00:30 AM at 10:01:30 AM) - then rapidapi would return the close price for the requested interval (ending at 10:01:00 AM) and it might fire the dispute even if both bots are run honestly.

Hence the only feasible method to have deterministic price resolution (independent on when the bot queries the API) is to use close price of the previous 1 minute interval (e.g. request price for timestamp 10:00:30 AM should be resolved with close price for the interval opening at 09:59:00 AM). In case of request being made during trading hours you should also filter out the last returned real-time price since that could be changing every time the request is made).

Also you should amend the the implementation section in the `uSPAC10` price identifier UMIP to reflect this same logic (using close price of previous 1 minute interval). Otherwise this could be rightfully disputed at the DVM escalation.

Reply:
> So I think the correct decision would be to always use _closing_ price for the interval, as it is currently implemented

Due to reasons in my previous comment it would be safer to use _close_ price for the previous interval.

Reply:
> In case the liquidation bot requests price within the same 1 minute interval as the current time (e.g. requesting price for timestamp 10:00:30 AM at 10:00:45 AM) there would not yet be the close price available for the current 1 minute interval.

As far as I can see, liquidation bot never calls `getHistoricalPrice`, it only calls `getCurrentPrice` which is implemented by using different call to RapidAPI. So the liquidation bot never operates with the notion of the exact price for some specific moment in time. It corresponds to the design of EMP smart contract. The function `createLiquidation` in EMP contract does not receive timestamp as an argument, the timestamp is taken from the block time. If we take into account block mining latency, then we can conclude that entire system starting from smart contract is not designed for 1min granularity.

In case you describe, when liquidation bot asks for the price in the middle of the interval, RapidAPI `Quotes` call will return the current price, which probably will be closer to the __current__ interval close price (which will be used by disputer), than __previous__ interval close time.

So I don't see any advantage in using __previous__ interval close price. But it has some difficulties. The major one is that the interval open price does not always equal to previous interval close price.

Another difficulty is that we will have two different cases. The first case is when the timestamp is in the middle of the day, the second case is when the timestamp is after the end of the day. In the last case we still have to use closing price for the current interval (which starts at 15:59:00). We have no easy way in code to decide which case do we have. We will have to either hardcode exchange closing time or implement some heuristic, like if the last interval time was more than 1 minute ago, then the day is closed.